Sources/Protocols/Designable/GradientDesignable.swift

```diff
@@ -6,145 +6,153 @@
     let gradientLayer = makeGradientLayer()
-    gradientLayer.colors = [startColor.cgColor, endColor.cgColor]
-    gradientLayer.configurePoints(with: startPoint)
+    gradientLayer?.frame = navBarFrame
+    gradientLayer?.cornerRadius = layer.cornerRadius
 
-    let image = gradientLayer.makeImage()
+    let image = gradientLayer?.makeImage()
     setBackgroundImage(image, for: .default)
   }
 
-  fileprivate func makeGradientLayer() -> CAGradientLayer {
-    let gradientLayer: CAGradientLayer = CAGradientLayer()
+}
 
-    let statusBarHeight: CGFloat = 64 // UIApplication.shared.statusBarFrame.height
-    let navBarFrame = CGRect(x: 0, y: 0, width: bounds.width, height: statusBarHeight)
-    gradientLayer.frame = navBarFrame
+// MARK: - Layers
 
-    gradientLayer.cornerRadius = layer.cornerRadius
-    return gradientLayer
+extension CALayer {
+
+  fileprivate func makeImage() -> UIImage? {
+    UIGraphicsBeginImageContext(frame.size)
+    render(in: UIGraphicsGetCurrentContext()!)
+    let outputImage = UIGraphicsGetImageFromCurrentImageContext()
+    UIGraphicsEndImageContext()
+    return outputImage
   }
 
 }
 
-private class PrivateGradientView: UIView {
+// MARK: - Gradients maker
 
-  init(frame: CGRect, gradientLayer: CAGradientLayer) {
-    super.init(frame: frame)
-    layer.insertSublayer(gradientLayer, at: 0)
-    autoresizingMask = [.flexibleWidth, .flexibleHeight]
+extension GradientDesignable {
+
+  private func makeGradientLayer() -> CALayer? {
+    guard let colors = gradientColors() else {
+      return nil
+    }
+
+    switch gradientMode {
+    case .linear:
+      return makeLinearLayer(colors: colors)
+    case .radial:
+      return makeRadialLayer(colors: colors)
+    }
   }
 
-  required init?(coder aDecoder: NSCoder) {
-    fatalError("init(coder:) has not been implemented")
+  private func gradientColors() -> [CGColor]? {
+    if let startColor = startColor, let endColor = endColor {
+      return [startColor.cgColor, endColor.cgColor]
+    } else if let startColor = predefinedGradient?.colors.start, let endColor = predefinedGradient?.colors.end {
+      return [startColor.cgColor, endColor.cgColor]
+    }
+    return nil
   }
 
-  override class var layerClass: AnyClass {
-    return CAGradientLayer.self
+  private func makeLinearLayer(colors: [CGColor]) -> CAGradientLayer {
+    let layer = CAGradientLayer()
+    layer.colors = colors
+
+    let points = layerPoints()
+    layer.startPoint = points.0
+    layer.endPoint = points.1
+    return layer
   }
 
-}
+  private func makeRadialLayer(colors: [CGColor]) -> RadialGradientLayer {
+    let layer = RadialGradientLayer()
+    layer.colors = colors
 
-fileprivate extension CAGradientLayer {
+    let points = layerPoints()
+    layer.startPoint = points.0
+    layer.endPoint = points.1
+    return layer
+  }
 
-  func configurePoints(with startPoint: GradientStartPoint) {
+  private func layerPoints() -> (CGPoint, CGPoint) {
```

Comment:
Could we keep the name `configurePoints`? It's a lot more readable imo

Reply:
I’d like to drop the `comfigure` because it doesn’t change the layer anymore, it will simply returns the points. Beside that, I’m completely open for another name!

Reply:
Maybe `gradientPonts`?

Reply:
Okay that makes sense. `gradientPoints` should work 😄